labs/lab03.md

```diff
@@ -0,0 +1,52 @@
+
+**Code coverage:** Coverage is the amount of code that has been executed by the tests. There are several coverage metrics, but one common metric line coverage is simply the percentage of lines (or statements) that have been executed by your unit tests. 
+
+## Deliverables
+
+- [ ] Write tests for LinkedIntQueue in TypeScript and ArrayIntQueue in Java and execute them from within the IDE.
+- [ ] Achieve 100% line coverage for ArrayIntQueue and show coverage results in the IDE or as HTML report. 
+- [ ] Compare structural testing vs specification testing and analyze tradeoffs. 
+
+
+## Instructions
+
+### Setup
+Fork and clone the repo from https://github.com/CMU-17-214/f23-rec03.
+The code is set up with JUnit and ts-jest. You can use your IDE to open the folder of each language and run the existing test with the command in the readme.txt. You can also see Appendix 1 for how to run tests in IDE. 
+
+
+### Testing
+Write additional tests for code we've provided and use a code coverage tool to measure the coverage of your unit tests. 
+
+For LinkedIntQueue you will need to use specification-based testing, because the source code for LinkedIntQueue is not provided. Your tests will be based on the specified functionality of the IntQueue interface rather than the code of the underlying LinkedIntQueue implementation. 
+1. Open TypeScript folder with IDE. Read the comments for the IntQueue interface in queue.js file, and familiarize yourself with each method. 
+2. Run queue.test.ts file(see readme.txt for instructions). Write more test cases to finish the specification testing for LinkedIntQueue. See Appendix 2 for notation.
+
+For ArrayIntQueue use structural testing. Specifically test underlying implementation of the ArrayIntQueue class, not just test against the IntQueue specification. 
+1. Open Java folder with IDE. Run IntQueueTest class in the IDE (see readme.txt). Change the mQueue variable to ArrayIntQueue and run testing again. Use your IDE to see the code coverage result. See appendix 3 for coverage tooling. 
+2. Try to find 2 bugs in ArrayIntQueue by running your test. Fix each bug that was found. 
+3. Write more test cases to achieve 100% code coverage. Hint: you also need to write tests for ensureCapacity. It is fine if you don’t have time to finish it during the class.
+
+
+## Best testing practices
+You were previously encouraged to read this post on unit testing best practices: https://www.testim.io/blog/unit-testing-best-practices/
+
+Consider the code in *sortArr*. It contains **a single test for a method called sortArr**. Go through the best practices discussed in the article above and identify which ones are violated. Furthermore, discuss how you would improve the test(s) to mitigate these concerns. You can but are not required to actual change the code.
```

Comment:
this is not currently a deliverable. I could be. or we could just remove this part. I don't like having extra instructions that do not go toward a deliverable. right now it's just some bonus, but it's not obious that it's not required.